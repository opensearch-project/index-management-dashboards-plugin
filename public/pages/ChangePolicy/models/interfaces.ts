/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Policy } from "../../../../models/interfaces";

export interface PolicyOption {
  label: string;
  value: Policy;
}
