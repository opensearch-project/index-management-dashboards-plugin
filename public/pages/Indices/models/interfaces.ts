/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { Policy } from "../../../../models/interfaces";
import { ManagedCatIndex } from "../../../../server/models/interfaces";

export interface PolicyOption {
  label: string;
  policy?: Policy;
}

export interface IndicesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof ManagedCatIndex;
  sortDirection: Direction;
  showDataStreams: boolean;
}
