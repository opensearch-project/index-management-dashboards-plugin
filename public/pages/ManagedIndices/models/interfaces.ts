/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { ManagedIndexItem } from "../../../../models/interfaces";

export interface ManagedIndicesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof ManagedIndexItem;
  sortDirection: Direction;
  showDataStreams: boolean;
}
