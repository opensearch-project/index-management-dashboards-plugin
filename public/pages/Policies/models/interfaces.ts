/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";

export interface PolicyItem {
  id: string;
  seqNo: number;
  primaryTerm: number;
  policy: object; // only dumped to view as JSON as of now, don't need to type
}

export interface PoliciesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof PolicyItem;
  sortDirection: Direction;
}
