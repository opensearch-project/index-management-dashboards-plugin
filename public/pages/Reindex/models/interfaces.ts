/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Conflicts } from "elasticsearch";

export interface ReindexRequest {
  waitForCompletion: boolean;
  slices?: number | string;
  maxDocs?: number;
  body: {
    conflicts?: Conflicts;
    source: {
      index: string;
      [key: string]: any;
    };
    dest: {
      index: string;
      pipeline?: string;
      op_type?: string;
    };
  };
  script?: {
    source: string;
    lang: string;
  };
}
export interface ReindexResponse {
  task: string;
}

export interface IndexSelectItem {
  status?: string;
  health?: string;
  isIndex?: boolean;
  isDataStream?: boolean;
  isAlias?: boolean;
  indices?: string[];
  writingIndex?: string;
}
