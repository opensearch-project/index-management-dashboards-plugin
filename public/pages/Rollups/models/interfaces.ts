/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { DocumentRollup } from "../../../../models/interfaces";

export interface RollupQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof DocumentRollup;
  sortDirection: Direction;
}
