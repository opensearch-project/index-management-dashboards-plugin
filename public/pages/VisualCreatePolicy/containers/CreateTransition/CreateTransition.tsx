/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component } from "react";
import { EuiText, EuiLink, EuiIcon, EuiFlyoutBody, EuiFlyoutFooter, EuiTitle, EuiFormRow, EuiSpacer, EuiComboBox } from "@elastic/eui";
import { Transition as ITransition, UITransition } from "../../../../../models/interfaces";
import FlyoutFooter from "../../components/FlyoutFooter";
import EuiFormCustomLabel from "../../components/EuiFormCustomLabel";
import { makeId } from "../../../../utils/helpers";
import Transition from "../../components/Transition";
import { TRANSITION_DOCUMENTATION_URL } from "../../../../utils/constants";

interface CreateTransitionProps {
  editTransition: UITransition | null;
  onCloseCreateTransition: () => void;
  onClickSaveTransition: (uiTransition: UITransition) => void;
  stateOptions: string[];
}

interface CreateTransitionState {
  uiTransition: UITransition | null;
}

export default class CreateTransition extends Component<CreateTransitionProps, CreateTransitionState> {
  constructor(props: CreateTransitionProps) {
    super(props);

    let uiTransition = props.editTransition;

    if (!uiTransition) {
      uiTransition = {
        transition: {
          state_name: "",
        },
        id: makeId(),
      };
    }

    this.state = { uiTransition };
  }

  onChange = (selectedOptions: { label: string }[]) => {
    // We should only get back either 0 or 1 options.
    const destinationState = selectedOptions.pop()?.label || "";
    const { uiTransition } = this.state;
    if (uiTransition == null) return;
    const newTransition: ITransition = { ...uiTransition.transition, state_name: destinationState };
    const newUiTransition = { ...uiTransition, transition: newTransition };
    this.setState({ uiTransition: newUiTransition });
  };

  onCreateOption = (searchValue: string) => {
    const normalizedSearchValue = searchValue.trim().toLowerCase();

    if (!normalizedSearchValue) {
      return;
    }

    const { uiTransition } = this.state;
    if (uiTransition == null) return;
    const newTransition: ITransition = { ...uiTransition.transition, state_name: normalizedSearchValue };
    const newUiTransition = { ...uiTransition, transition: newTransition };
    this.setState({ uiTransition: newUiTransition });
  };

  onChangeTransition = (uiTransition: UITransition) => {
    this.setState({ uiTransition });
  };

  onClickSaveTransition = () => {
    const { uiTransition } = this.state;
    if (!uiTransition) return;
    this.props.onClickSaveTransition(uiTransition);
  };

  render() {
    const { editTransition, onCloseCreateTransition, stateOptions } = this.props;
    const { uiTransition } = this.state;
    let title = "Add transition";
    if (!!editTransition) title = "Edit transition";
    return (
      <>
        <EuiFlyoutBody>
          <EuiTitle size="s">
            <h3>{title}</h3>
          </EuiTitle>

          <EuiText size="xs" style={{ fontWeight: 200 }}>
            Transitions define the conditions that need to be met for a state to change. After all actions in the current state are
            completed, the policy starts checking the conditions for transitions.{" "}
            <EuiLink href={TRANSITION_DOCUMENTATION_URL} target="_blank" rel="noopener noreferrer">
              Learn more
            </EuiLink>
          </EuiText>

          <EuiSpacer />

          <EuiFormCustomLabel
            title="Destination state"
            helpText="If entering a state that does not exist yet then you must create it before creating the policy."
          />

          <EuiFormRow fullWidth isInvalid={false} error={null}>
            <EuiComboBox
              fullWidth
              isClearable={false}
              placeholder="Select a single option"
              singleSelection={{ asPlainText: true }}
              options={stateOptions.map((state) => ({ label: state }))}
              selectedOptions={[{ label: uiTransition?.transition.state_name || "" }]}
              onChange={this.onChange}
              onCreateOption={this.onCreateOption}
              customOptionText="Add {searchValue} state"
            />
          </EuiFormRow>

          <EuiSpacer />

          {!!uiTransition && <Transition uiTransition={uiTransition} onChangeTransition={this.onChangeTransition} />}
        </EuiFlyoutBody>
        <EuiFlyoutFooter>
          <FlyoutFooter
            edit={!!editTransition}
            action="transition"
            onClickCancel={onCloseCreateTransition}
            onClickAction={this.onClickSaveTransition}
          />
        </EuiFlyoutFooter>
      </>
    );
  }
}
