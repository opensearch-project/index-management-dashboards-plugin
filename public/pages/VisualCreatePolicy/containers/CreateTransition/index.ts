/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import CreateTransition from "./CreateTransition";

export default CreateTransition;
