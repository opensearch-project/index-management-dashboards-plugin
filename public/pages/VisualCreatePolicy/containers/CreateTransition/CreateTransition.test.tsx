/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import CreateTransition from "./CreateTransition";
import { UITransition } from "../../../../../models/interfaces";

describe("<CreateTransition /> spec", () => {
  it("renders the component", () => {
    const transition: UITransition = { id: "some_id", transition: { state_name: "hot", conditions: { min_index_age: "30d" } } };
    const { container } = render(
      <CreateTransition
        editTransition={transition}
        onCloseCreateTransition={() => {}}
        onClickSaveTransition={() => {}}
        stateOptions={["hot", "warm", "cold"]}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
