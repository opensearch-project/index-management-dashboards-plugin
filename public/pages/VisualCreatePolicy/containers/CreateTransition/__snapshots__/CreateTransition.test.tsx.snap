// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateTransition /> spec renders the component 1`] = `
<div
  class="euiFlyoutBody"
>
  <div
    class="euiFlyoutBody__overflow"
    tabindex="0"
  >
    <div
      class="euiFlyoutBody__overflowContent"
    >
      <h3
        class="euiTitle euiTitle--small"
      >
        Edit transition
      </h3>
      <div
        class="euiText euiText--extraSmall"
        style="font-weight: 200;"
      >
        Transitions define the conditions that need to be met for a state to change. After all actions in the current state are completed, the policy starts checking the conditions for transitions.
         
        <a
          class="euiLink euiLink--primary"
          href="https://opensearch.org/docs/im-plugin/ism/policies/#transitions"
          rel="noopener noreferrer"
          target="_blank"
        >
          Learn more
          EuiIconMock
          <span
            class="euiScreenReaderOnly"
          >
            (opens in a new tab or window)
          </span>
        </a>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiText euiText--medium"
        style="margin-bottom: 5px;"
      >
        <h5
          class="euiFormLabel "
          style="margin-bottom: 2px;"
        >
          Destination state
        </h5>
        <p>
           
          <span
            style="font-weight: 200; font-size: 12px;"
          >
            If entering a state that does not exist yet then you must create it before creating the policy.
             
          </span>
        </p>
      </div>
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            class="euiComboBox euiComboBox--fullWidth"
            role="combobox"
          >
            <div
              class="euiFormControlLayout euiFormControlLayout--fullWidth"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiComboBox__inputWrap euiComboBox__inputWrap--fullWidth euiComboBox__inputWrap--noWrap"
                  data-test-subj="comboBoxInput"
                  tabindex="-1"
                >
                  <span
                    class="euiComboBoxPill euiComboBoxPill--plainText"
                  >
                    hot
                  </span>
                  <div
                    class="euiComboBox__input"
                    style="font-size: 14px; display: inline-block;"
                  >
                    <input
                      aria-controls=""
                      data-test-subj="comboBoxSearchInput"
                      id="some_html_id"
                      role="textbox"
                      style="box-sizing: content-box; width: 2px;"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <button
                    aria-label="Open list of options"
                    class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                    data-test-subj="comboBoxToggleListButton"
                    type="button"
                  >
                    EuiIconMock
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiText euiText--medium"
        style="margin-bottom: 5px;"
      >
        <h5
          class="euiFormLabel "
          style="margin-bottom: 2px;"
        >
          Condition
        </h5>
        <p>
           
          <span
            style="font-weight: 200; font-size: 12px;"
          >
            Specify the condition needed to be met to transition to the destination state.
             
          </span>
        </p>
      </div>
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--fullWidth"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <select
                class="euiSelect euiSelect--fullWidth"
                data-test-subj="create-state-action-type"
                id="some_html_id"
                style="text-transform: capitalize;"
              >
                <option
                  value="none"
                >
                  No condition
                </option>
                <option
                  value="min_index_age"
                >
                  Minimum index age
                </option>
                <option
                  value="min_doc_count"
                >
                  Minimum doc count
                </option>
                <option
                  value="min_size"
                >
                  Minimum size
                </option>
                <option
                  value="min_rollover_age"
                >
                  Minimum rollover age
                </option>
                <option
                  value="cron"
                >
                  Cron expression
                </option>
              </select>
              <div
                class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  EuiIconMock
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiText euiText--medium"
        style="margin-bottom: 5px;"
      >
        <h5
          class="euiFormLabel "
          style="margin-bottom: 2px;"
        >
          Minimum index age
        </h5>
        <p>
           
          <span
            style="font-weight: 200; font-size: 12px;"
          >
            The minimum age required to transition to the next state.
             
          </span>
        </p>
      </div>
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--fullWidth"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                class="euiFieldText euiFieldText--fullWidth"
                data-test-subj="transition-render-conditions-min-index-age"
                id="some_html_id"
                type="text"
                value="30d"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
